[PROPOSAL] Add summary, parent and kind as optional fields for tags (#4063)
OAI/OpenAPI-Specification: proposals/2024-09-01-Tags-Improvement.md

# Tags Improvement

## Metadata

|Tag |Value |
|---- | ---------------- |
|Proposal |[2024-09-01-Tags-Improvement](https://github.com/OAI/OpenAPI-Specification/tree/main/proposals/{2024-09-01-Tags-Improvement.md})|
|Authors|[Lorna Mitchell](https://github.com/lornajane)|
|Review Manager | TBD |
|Status |Proposal|
|Implementations | |

> |Issues | [1367](https://github.com/OAI/OpenAPI-Specification/issues/1367), [{Iss}](https://github.com/OAI/OpenAPI-Specification/issues/{Iss}), |

Reviewer: ```suggestion
|Issues | [1367](https://github.com/OAI/OpenAPI-Specification/issues/1367) |
```

Author: Ah, no I wanted to link something else here. I'll fix it

|Previous Revisions | None, but see [Moonwalk discussion](https://github.com/OAI/sig-moonwalk/discussions/67)

## Change Log

|Date |Responsible Party |Description |
|---- | ---------------- | ---------- |
| 2024-09-01 | @lornajane | Initial draft |

## Introduction

Evolve the existing `tags` implementation to (optionally) support more use cases, such as giving the tags some grouping/relationships and adding more metadata.

## Motivation

The tags feature hasn't changed since the spec was still called Swagger, but it's one of the most-extended aspects of OpenAPI with most tools supporting some sort of grouping or additional metadata. One motivation for this proposal is to "pave the cowpath" and formalise some of the patterns from those extensions as part of the specification.

> The existing tags implementation is also quite limiting, so users/tools feel they "cannot" use tags, because they are so widely implemented for documentation navigation and required exactly one tag per operation, to be used only for this single purpose or to be opted out using extensions such as `x-traitTag`. The result is that we see lots of extensions that add arbitrary metadata to operations, some of them even become part of the specification officially (looking at you, `deprecated: true`) or being so widely used that we should probably adopt them (`x-internal`, `x-experimental`). The specification does have a way of tagging operations, but it's too limited and the result is that everything has to be added as an extension field.

Reviewer: "required exactly one tag per operation" - is that an example for (ab)use of tags?

Author: It's a very common pattern of what downstream tools expect

Reviewer: Well...

> On a personal note, I work for a tool vendor and was proposing these changes internally; but the problems affect all OpenAPI users so I brought it to the main project.

Reviewer 2: _Thank you!!!_  We need more of this kind of tool to spec feedback loop.

### Supporting evidence

> There are several examples "in the wild" of where a better tags implementation would have helped. Here is a selection of publicly-accessible examples here to illustrate some of the problems this proposal could help with:

Reviewer: ```suggestion
There are several examples "in the wild" of where a better tags implementation would have helped. Here is a selection of publicly-accessible examples to illustrate some of the problems this proposal could help with:
```

- Various tag-alike additions exist, sometimes called "badges" or similar; I'd include extensions such as [`x-internal`](https://redocly.com/docs/cli/guides/hide-apis/#step-1-add-x-internal-to-the-api-description) as a tag-alike since they could be tags if more than one tag (or tag type) could be applied.
- Additional display metadata is also in active use in the wild, see [`x-displayName`](https://redocly.com/docs/api-reference-docs/specification-extensions/x-display-name) and this [OpenAPI specification issue](https://github.com/OAI/OpenAPI-Specification/issues/2843)

## Proposed solution

Originally proposed in a [Moonwalk discussion](https://github.com/OAI/sig-moonwalk/discussions/67), I am proposing three backwards-compatible additions to the existing tags feature:

* Tags get a `summary` alongside `name` and `description`. In keeping with our existing practices: name is the identifier, summary is the short display content, and description is available in contexts where more information is appropriate.
* A `parent` field is added to support a hierarchy without adding either separators or a new data type. Your tag can belong to another tag.

> * A `kind` field to explain which family of tags this tag belongs to (previously proposed as `type`). We'd expecting these to be `nav`, `badge`, `internal` and probably some other things that other tooling types would find useful.

Reviewer 2: (Warning: bikeshedding ahead, feel free to ignore). As "kind" has two meanings would "category" be more clear for non-native speakers?  I agree on avoiding "type" as it usually means data type.

Reviewer: "Category" has its own heap of baggage, I find "kind" less overloaded and easier to understand here.

Reviewer 2: @ralfhandl I'm fine with not using "category" but I am curious as to what baggage it has in this context?

Author: I don't love or hate any of the options

Reviewer: @handrews It's probably only my experience with "Foo category" being a coarse-grained grouping of Foos, and "Foo type" being a refinement of "Foo category" 😄

---|:---:|---
<a name="tagName"></a>name | `string` | **REQUIRED**. The name of the tag. Use this value in the `tags` array of an Operation.
<a name="tagSummary"></a>summary | `string` | A short summary of the tag, used for display purposes.
<a name="tagDescription"></a>description | `string` | A description for the tag. [CommonMark syntax](https://spec.commonmark.org/) MAY be used for rich text representation.
<a name="tagExternalDocs"></a>externalDocs | [External Documentation Object](#external-documentation-object) | Additional external documentation for this tag.
<a name="tagParent"></a>parent | `string` | The `name` of a tag that this tags is nested under. The named tag MUST exist in the API description, and circular references between parent and child tags MUST NOT be used.
<a name="tagKind"></a>kind | `string` | A machine-readable string to categorize what sort of tag it is. Common uses are `nav` for Navigation, `badge` for badges, `internal` for internal APIs, but any string value can be used. A registry of known values is available.

This object MAY be extended with [Specification Extensions](#specification-extensions).

##### Tag Object Example

```json

> {
> 	"name": "account-updates",
>     "summary": "Account Updates",
> 	"description": "Account update operations",
>     "type": "nav"
> },
> {
> 	"name": "partner",
>     "summary": "Partner",
> 	"description": "Operations available to the partners network",
>     "parent": "external",
>     "type": "audience"
> }

Reviewer: ```suggestion
{
	"name": "account-updates",
  "summary": "Account Updates",
	"description": "Account update operations",
  "kind": "nav"
},
{
	"name": "partner",
  "summary": "Partner",
	"description": "Operations available to the partners network",
  "parent": "external",
  "kind": "audience"
},
{
	"name": "external",
  "summary": "External",
	"description": "Operations available to external consumers",
  "kind": "audience"
}
```

```

```yaml
- name: account-updates
  summary: Account Updates
  description: Account update operations
  type: nav

- name: partner
  summary: Partner
  description: Operations available to the partners network

>   parent: external

Reviewer: Can we add the `external` tag?

Reviewer 2: @ralfhandl I view this text as just giving us the general idea of how the proposal would be implemented.  We can nail down the spec wording and examples when the spec changes are made.

(we should probably figure out what the balance is between the proposals and the specification text, idk, maybe folks want to get the wording right here, in which case I'm happy to defer to that approach)

Reviewer: Depending on how long this proposal will sit here without being incorporated into a specification version I'd like to avoid the impression that the proposal allows/advertises the use of dangling "parent" values by adding just one more tag object to the example and make it self-contained.

Reviewer 2: @ralfhandl I see- that makes sense.

> - name: account-updates
>   summary: Account Updates
>   description: Account update operations
>   type: nav
> 
> - name: partner
>   summary: Partner
>   description: Operations available to the partners network
>   parent: external
>   type: audience

Reviewer: ```suggestion
- name: account-updates
  summary: Account Updates
  description: Account update operations
  kind: nav

- name: partner
  summary: Partner
  description: Operations available to the partners network
  parent: external
  kind: audience

- name: external
  summary: External
  description: Operations available to external consumers
  kind: audience
```
